[3006.x] Don't needlessly refresh fileserver backends during Pillar rendering

terminalmage opened the pull request:
### What does this PR do?

This uses the opts key that masterless runs use to limit fileserver backend refreshes to a single refresh. Adding it to the master opts ensures that FSChan instances created on the master do not trigger fileserver backend refreshes.

### What issues does this PR fix or reference?

Fixes #65990

### Previous Behavior

Fileserver backends are refreshed when the master instantiates an `FSClient` instance.

### New Behavior

Fileserver backends are no longer refreshed.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
No

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

terminalmage committed:
Don't needlessly refresh fileserver backends during Pillar rendering

This uses the opts key that masterless runs use to limit fileserver
backend refreshes to a single refresh. Adding it to the master opts
ensures that FSChan instances created on the master do not trigger
fileserver backend refreshes.

terminalmage committed:
Rename test

terminalmage committed:
_get_defaults can be a staticmethod

dmurphy18 reviewed (changes requested):
The tests case needs to be under tests/pytest and using pytest to test with.
Tests under tests/unit are considered unittest and will eventually be removed.

All tests PR's need tests now written using pytest, are they will be rejected.

terminalmage committed:
Move tests to pytests dir and reformat them

terminalmage commented:
> The tests case needs to be under tests/pytest and using pytest to test with. Tests under tests/unit are considered unittest and will eventually be removed.
> 
> All tests PR's need tests now written using pytest, are they will be rejected.

I brought this up with @dmurphy18 in Slack, but this is directly contradicted by the following docs pages:

- https://docs.saltproject.io/en/latest/topics/development/contributing.html
- https://docs.saltproject.io/en/latest/topics/tutorials/writing_tests.html

Additionally, the location where tests should be added is _not_ covered by the recent changes to [Project Management](https://docs.saltproject.io/en/master/topics/maintenance-policy.html). Both the above two pages should arguably be linked from this page.

With two distinct sets of tests in the repo, it's a must to get this documentation updated to reflect the current policies of the project. Failure to do so increases the burden placed on the core team, as well as the contributors. 

I've moved the tests over. I'm also available to assist in the process of updating the docs so that they more accurately reflect reality, provided I can get better clarity on the changes to these policies.

s0undt3ch committed:
Switch to existing test fixtures

terminalmage committed:
Rename test functions to specify that they are testing FSClient

The original draft of these tests was class-based, with the class name
describing that the tests were for FSClient. With the class removed,
this should be in the test name.

s0undt3ch commented on salt/config/__init__.py:
```diff
@@ -3952,6 +3952,14 @@ def apply_master_config(overrides=None, defaults=None):
+    # in its maintenance thread. Refreshing during Pillar compilation slows
+    # down Pillar considerably (even to the point of timeout) when there are
+    # many gitfs remotes.
+    opts["__fs_update"] = True
```
I don't like the fact that we are using `opts` to tweak behavior "down the pipeline".
With that said, you're not introducing that behavior, it's been long present in Salt.

terminalmage replied:
Given that fileclients are generated automatically when needed, and that only the opts are passed in (see `salt.fileclient.get_file_client()`), I don't see a better solution than to use the opts. If you have a better idea, I'm interested. However, any changes to how fileclients are spawned would have impacts to a number of other systems, which is why I didn't try to reinvent the wheel when I made my bugfix.

s0undt3ch replied:
> Given that fileclients are generated automatically when needed, and that only the opts are passed in (see `salt.fileclient.get_file_client()`), I don't see a better solution than to use the opts. If you have a better idea, I'm interested. However, any changes to how fileclients are spawned would have impacts to a number of other systems, which is why I didn't try to reinvent the wheel when I made my bugfix.

That also why I did not suggest an alternate approach at this moment.

s0undt3ch reviewed (dismissed):
I updated your tests to use existing test fixtures and rebased the PR against 3006.x

terminalmage commented:
You broke the tests, I'm fixing them

terminalmage committed:
Fix tests

terminalmage committed:
Don't monkeypatch the fixture contents

terminalmage commented:
@s0undt3ch OK, tests are fixed now.

You can't just use the opts dict, because the way salt actually works when it stands up a master is it loads the config using `salt.config.master_config()`, passing it the path to the config file. Within that function it runs `salt.config.apply_master_config()`, passing in the defaults and anything loaded from the config file path passed to `salt.config.master_config()`.

Since we aren't loading from a config file, running `salt.config.apply_master_config()` and passing in our own overrides (and using the fixture data as the defaults) is a close enough approximation for testing purposes.


Note that I also removed the monkeypatching of the fixture dictionaries. I forgot what project the fixtures come from, but I was worried that adding a key to that dictionary would pollute other tests if minion_opts/master_opts weren't function-scoped.

s0undt3ch commented:
> You broke the tests, I'm fixing them

Hmm, funny, they did pass locally.

s0undt3ch reviewed (dismissed):
I'm good with the latest changes

terminalmage commented on salt/config/__init__.py:
```diff
@@ -3948,6 +3948,14 @@ def apply_master_config(overrides=None, defaults=None):
     opts = defaults.copy()
     opts["__role"] = "master"
+
+    # Suppress fileserver update in FSChan, for LocalClient instances generated
```
@s0undt3ch @twangboy I know you've both already approved the PR, but this line should say FSClient instead of LocalClient. Mind if I change it before merging? Feel free to also just make this change yourself.

s0undt3ch replied:
Sure, go ahead a commit the change.

terminalmage committed:
Change LocalClient -> FSClient

The LocalClient name has been deprecated for some time and now exists
only as an alias.
